feat: Introduce `wallpapers` module

fiftydinar opened the pull request:
This module will automatically copy wallpapers into system directory for all image variants.

If image is Gnome-based, it can additionally support advanced features:

- Setting all wallpapers to appear in Gnome settings, Appearance section.
- Including specified light+dark wallpapers
- Setting default wallpaper, including light+dark variant
- Setting various image scaling options, globally &/or per-wallpaper

Gnome additional support, which is not planned or not yet implemented:

- Solid & Gradient color wallpapers (also acts as a scaling filler in non-wallpaper area for normal wallpapers)
- Dynamic wallpapers (can be based on various time conditions)
- ... other things which I don't know or can't remember

Other DEs support:
- Let me know

I tested this thoroughly & it works great!

I made this module to be as easy to maintain as possible in the code.
F.e, if some scaling option gets depreciated, it is very easy to remove it in the code, inside `scaling_options` variable.
I also categorized the code & explained as much as possible.

Only thing that can be possibly adjusted in this PR is documentation & maybe some other minor things.

Image format support & file-size is mentioned as something that should be checked before applying.
Image format support is dynamically changing, so I'm not sure is it worth to implement this.
File-size check can be easily done, to prevent image bloat.

External wallpaper source is complex as mentioned, so we would probably not support that. Plus we don't know which source we would use.
This module itself with current functionality is the most complex bash module made for BlueBuild as of now, so we have to think about implementing additional functionality carefully.

Related issue:
https://github.com/ublue-os/bling/issues/130

fiftydinar committed:
feat: Introduce `wallpapers` module

This module will automatically copy wallpapers into system directory for all image variants.

If image is Gnome-based, it can additionally support advanced features:

- Setting all wallpapers to appear in Gnome settings, Appearance section.
- Including specified light+dark wallpapers
- Setting default wallpaper, including light+dark variant
- Setting various image scaling options, globally &/or per-wallpaper

Gnome additional support, which is not planned or not yet implemented:

- Setting different colors & gradients for background scaling filler (default is black).
- Dynamic wallpapers (can be based on various time conditions)
- ... other things which I don't know or can't remember

Other DEs support:
- Let me know

I tested this thoroughly & it works great!

I made this module to be as easy to maintain as possible in the code. F.e, if some scaling option gets depreciated, it is very easy to remove it in the code, inside `scaling_options` variable.
I also categorized the code & explained as much as possible.

Only thing that can be possibly adjusted in this PR is documentation & maybe some other minor things.

xynydev commented:
Thanks! Reviewing this is not a priority right now. I skimmed the source code just now, no glaring issues.  
A couple notes though:
- Why is the default wallpaper configured as an array?
- What does `zoom: all` mean in practice when all specified wallpapers use a different scaling option (except the default)?
- Why use the `light + dark` syntax which might encounter unforeseen parsing errors, instead of just:
```yaml
- filename-any.jpg
- dark: filename-dark.jpg
  light: filename-light.jpg
```

The config format feels a bit weird overall. Here's just a sketch for something different:
```yaml
type: wallpapers
default-scaling: wallpaper
wallpapers:
   - file: filename-tile.jpg
   - dark: filename-center-dark.jpg
     light: filename-center-light.jpg
     scaling: centered
     default: true
```
or if the defaults shouldn't work like that:
```yaml
type: wallpapers
default-scaling: wallpaper
wallpapers:
   - file: filename-tile.jpg
default-wallpaper:
     dark: filename-center-dark.jpg
     light: filename-center-light.jpg
     scaling: centered
```

fiftydinar commented:
> Why is the default wallpaper configured as an array?

I can fix that, I set them as array just because of convenience (already available sanitize_file_names variable & such). 

> What does zoom: all mean in practice when all specified wallpapers use a different scaling option (except the default)?

Per-wallpaper scaling is always overwriting global option.
I can change the name of this option similar to  your example & note that in recipe too (currently noted in README only).

> Why use the light + dark syntax which might encounter unforeseen parsing errors

I use this format because current light+dark wallpaper is easier to manage & to notice compared to separating them.
F.e., when having good amount of those wallpapers, you would not need to jump on finding & removing combination 1 by 1 (light by dark), but you would just remove a pair.
I think that newlines are not a problem, except trailing newlines, which are removed (upstreamed in bluebuild-cli export too inside get_yaml_array function).
Another factor is the convenience of the user, where he can include wallpaper files with whitespaces, without need to rename them. I automatically convert those whitespaces to _ character.

1st recipe looks kinda cluttered to me, with default wallpaper, light+dark wallpaper, wallpaper & scaling together.

2nd recipe looks better, but still not my taste, as I prefer light+dark wallpaper as a current format.
I can maybe merge scaling into those, but I will have to think on how that will look.

fiftydinar committed:
Merge remote-tracking branch 'upstream/main' into wallpapers-module

fiftydinar committed:
Merge remote-tracking branch 'upstream/main' into wallpapers-module

fiftydinar committed:
fix(wallpapers): Unbound variable for MODULE_DIRECTORY on new templates

fiftydinar committed:
Merge remote-tracking branch 'upstream/main' into wallpapers-module

fiftydinar committed:
Merge remote-tracking branch 'upstream/main' into wallpapers-module

fiftydinar committed:
Merge remote-tracking branch 'upstream/main' into wallpapers-module

fiftydinar committed:
Merge branch 'main' into wallpapers-module

fiftydinar committed:
Merge branch 'main' into wallpapers-module

fiftydinar committed:
fix(wallpapers): Minor function format fix

fiftydinar committed:
chore(wallpapers): Improve code formatting style

This would fit with future code guidelines outlined here:

https://github.com/blue-build/modules/issues/157

fiftydinar committed:
fix: Quoted * as literal string, instead as an expansion

tulilirockz commented:
Guys maybe it would be interesting to have a "wallpaper-packs" option or something like that, because it would be awesome to add stuff like https://github.com/tulilirockz/ublue-os-wallpapers as easily as adding a single string like `- wallpaper-packs: - $URL`

xynydev commented:
@tulilirockz Could be cool. We've discussed adding different ways to fetch wallpapers. I think `git clone` is a more promising way than custom rpm repos, though.

In the meantime, I saw that I actually kept the ublue wallpaper bling installer, so you can go ahead and add back a script there to get the wallpapers from your rpm if you wish: https://github.com/blue-build/modules/blob/main/modules/bling/installers/ublue-os-wallpapers.sh

tulilirockz commented:
Sweet! Gonna make a PR adding them back up on the script :>

fiftydinar committed:
Merge branch 'main' into wallpapers-module

fiftydinar committed:
Revert "fix: Quoted * as literal string, instead as an expansion"

This reverts commit cc640b5ecba65b5eee8d5151a4170cfae7e71f38.

fiftydinar committed:
Revert "chore(wallpapers): Improve code formatting style"

This reverts commit 82c4abc19c1e08962a5693ad33e5e72a8b0c83e7.

fiftydinar committed:
Merge branch 'main' into wallpapers-module

fiftydinar commented:
I will mark this PR as draft, until it's investigated on how to support other DEs better.
And until module code guidelines are formed.

fiftydinar committed:
Merge remote-tracking branch 'upstream/main' into wallpapers-module

fiftydinar committed:
fix: Improve Gnome detection

Instead of parsing if session exists, check if `gnome-shell` command exists.

fiftydinar commented:
I will close this PR & make new revised one when it becomes better, as this is required:

- final agreement on how module recipe will look like
- this code requires refactor to make code much clearer, since it looks messy right now
- support other desktop environments for additional features
